generic-ec-zkp/src/dlog_eq.rs

```diff
@@ -0,0 +1,277 @@
+//!     pub_share: X,
+//!     base: H,
+//!     exp: Z,
+//! };
+//!
+//! // Prover chooses a private nonce
+//! let r = Scalar::random(&mut OsRng);
+//! // Prover proves in zero knowledge the equality of `H * x = Z`
+//! let proof = non_interactive::prove::<sha2::Sha256, E>(&"shared_state", &x, data, r);
+//! // The proof is sent to the Verifier
+//! send(proof);
+//!
+//! // Verifier checks that the proof is correct
+//! non_interactive::verify::<sha2::Sha256, E>(&"shared_state", data, proof)
+//!     .expect("Verification failed!");
+//! ```
+//!
+//! ## Interactive example
+//!
+//! ```rust
+//! # use generic_ec::{Curve, Scalar, SecretScalar, Point};
+//! # use generic_ec_zkp::dlog_eq::interactive;
+//! # use rand::rngs::OsRng;
+//! # use generic_ec::curves::Secp256k1 as E;
+//! # fn send<T>(_: T) {}
+//!
+//! // `x` is a secret discrete logarithm only known to prover
+//! let x = SecretScalar::<E>::random(&mut OsRng);
+//!
+//! // `X` is a public point corresponding to `x`
+//! let X = Point::generator() * &x;
+//! // `H` is some point known to both parties
+//! let H = Point::generator() * Scalar::random(&mut OsRng);
+//! // `Z` is the object of the proof
+//! let Z = H * &x;
+//!
+//! let data = interactive::Data {
+//!     pub_share: X,
+//!     base: H,
+//!     exp: Z,
+//! };
+//!
+//! // Prover chooses a nonce for itself
+//! let r = Scalar::random(&mut OsRng);
+//! // Prover commits to the data
+//! let commitment = interactive::commit_data(data, r);
+//! // Prover sends this commitment to the verifier
+//! send(commitment);
+//!
+//! // Verifier sends a challenge to the prover
+//! let challenge = Scalar::random(&mut OsRng);
+//! send(challenge);
+//!
+//! // Prover receives the challenge and computes the proof of equality `H * x = Z`
+//! let proof = interactive::prove(r, challenge, &x);
+//! // This proof is sent to the verifier
+//! send(proof);
+//!
+//! // Verifier checks that the proof is correct
+//! interactive::verify(data, commitment, challenge, proof)
+//!     .expect("Verification failed!");
+//! ```
+
+use generic_ec::{Curve, Point};
+
+/// Object of the proof
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "udigest", derive(udigest::Digestable), udigest(bound = ""))]
+pub struct Data<E: Curve> {
+    /// `X`, the value inside `log_G`
+    pub pub_share: Point<E>,
+    /// `H`, the base for the second logarithm
+    pub base: Point<E>,
+    /// `Z`, the value inside the second logarithm `log_H`
+    pub exp: Point<E>,
+}
+
+/// Functions for interactive protocol for the proof. See usage example in
+/// [`dlog_eq`](crate::dlog_eq)
+///
+/// Based on "Zero-Knowledge Proofs Notes", 2024 by Jorge L. Villar
+pub mod interactive {
+    use generic_ec::{Curve, Point, Scalar, SecretScalar};
+
+    /// Commitment to `H`, sent in the first round
+    pub type Commitment<E> = (Point<E>, Point<E>);
+    /// Challenge, sent to the Prover in the first round
+    pub type Challenge<E> = Scalar<E>;
```

Comment:
No function to generate a challenge. We usually add this function, even if all it does is sample a scalar